### PR TITLE
feat(spin/template-use)

### PR DESCRIPTION
**Pull request by linjmeyer**
[As discussed in issue 203](https://github.com/spinnaker/spin/issues/203)

**Usage:**

Basic usage:
`spin pt use my-template -a my-application -n my-pipeline`

Creating a pipeline using spin pipeline save:
`spin pt use my-template -a my-application -n my-pipeline | spin pipeline save`

Pipeline Template variables:
`spin pt use my-template -a my-application -n my-pipeline --set myvar=myvalue,myothervar=myothervalue`

Pipeline template variable files:
`spin pt use my-template -a my-application -n my-pipeline --values myfile.yaml,myotherfile.json`

Using `--set` and `--values` together is supported.  If both contain the same variable, set is used over value.

Other options:
* --tag for setting the tag
* --description or -d for setting the description of the pipeline
* --type for setting the template type (default front50/pipelineTemplate)
* --artifact-account for setting artifact account (default front50ArtifactCredentials)

First time playing with go and contributing to spin so feedback is very welcome

### CODE REVIEW

**Comment by milesmatthias** [comment]
Could we see an example with source files here? I forgot to ask a clarifying question in that discussion -- are the generated pipelines going to actually be using pipeline templates, or is this command rendering raw, basic pipelines (ie, pipelines without depending on the presence of MPT)?

Happy to contribute to a documentation PR for this over at https://github.com/spinnaker/spinnaker.github.io.

**Comment by linjmeyer** [comment]
@milesmatthias sorry forgot to mention it:  I didn't add the "fully hydrated" pipeline but still think its a good idea, just wanted to start with the simpler use case and add that feature after this one is accepted.

I'll add some example templates and corresponding commands this evening

**Comment by linjmeyer** [comment]
## Example with source files

### Save a template that expects a few variables
Variables are:
* judgeMessage: The message that should be displayed in the manual judgement step
* timeToWait: Number of seconds the wait stage should wait for

Template:
```json
{
  "schema": "v2", 
  "variables": [
    {
      "type": "int",
      "description": "The time a wait stage shall pauseth",
      "name": "timeToWait"
    },
    {
        "type": "string",
        "description": "The message to be displayed in the starting manual judgement step",
        "name": "judgeMessage"
    }
  ],
  "id": "spin-pt-use-example",
  "protect": false,
  "metadata": {
    "name": "Variable Wait",
    "description": "A demonstrative Wait Pipeline.",
    "owner": "example@example.com",
    "scopes": ["global"]
  },
  "pipeline": { 
    "lastModifiedBy": "anonymous",
    "updateTs": "0",
    "parameterConfig": [],
    "limitConcurrent": true,
    "keepWaitingPipelines": false,
    "description": "",
    "triggers": [],
    "notifications": [],
    "stages": [
        {
            "failPipeline": true,
            "instructions": "${ templateVariables.judgeMessage }",
            "isNew": true,
            "judgmentInputs": [],
            "name": "Manual Judgment",
            "notifications": [],
            "type": "manualJudgment",
            "refId": "judge1"
        },
        {
            "waitTime": "${ templateVariables.timeToWait }", 
            "name": "My Wait Stage",
            "type": "wait",
            "refId": "wait1",
            "requisiteStageRefIds": ["judge1"]
        }
    ]
  }
}
```

### Run spin pipeline template use to generate the pipeline

Set the variable timeToWait using `--set timeToWait=40`, set the Jenkins template variables using a key/value json file and `--values template-values.json`

template-values.json:
```json
{
    "judgeMessage": "Are you super duper sure you want to run this pipeline?"
}
```

The final command will look something like:

`spin pt use spin-pt-use-example --application your-spinnaker-app --name your-new-pipeline --set timeToWait=40 --values template-values.json > your-new-pipeline.json`

your-new-pipeline.json file contents _should_ look like this:

```json
{
 "application": "your-spinnaker-app",
 "description": "",
 "exclude": [],
 "name": "your-new-pipeline",
 "notifications": [],
 "parameters": [],
 "schema": "v2",
 "stages": [],
 "template": {
  "artifactAccount": "front50ArtifactCredentials",
  "reference": "spinnaker://spin-pt-use-example",
  "type": "front50/pipelineTemplate"
 },
 "triggers": [],
 "variables": {
  "judgeMessage": "Are you super duper sure you want to run this pipeline?",
  "timeToWait": "40"
 }
}
```

### Save and test the pipeline

Save the pipeline with `spin pi save --file your-new-pipeline.json`

Navigating to your pipeline and hitting configure you should see your variables:

image

No configuration issues/errors showing in the pipeline:

image

Finally running the pipeline you should see your variables in action:

image

image

**Comment by louisjimenez** [comment]
Thanks again for the great first time contribution @linjmeyer! I called out your feature at the [Spinnaker As Code SIG](https://docs.google.com/document/d/1qgtwod1WC9pGNLyq7j1pXDLCAhO3Norf3Afx1FP4Jag/edit#heading=h.h9ro0baz3lei) to get some more eyes and feedback on it. This is looking good so far, I should be able to get it merged tomorrow and released alongside 1.16  next week.

File: cmd/pipeline-template/use.go [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+	application     string
+	name            string
+	description     string
+	variables       map[string]string
+	templateType    string
+	artifactAccount string
+	variablesFiles  []string
+}
+
+var (
```
[the remark this answers is not in the document]
Comment (linjmeyer) [reply cmd/pipeline-template/use.go]: Not sure what you mean?

[the remark this answers is not in the document]
Comment (louisjimenez) [reply cmd/pipeline-template/use.go]: how about const instead of var?

[the remark this answers is not in the document]
Comment (linjmeyer) [reply cmd/pipeline-template/use.go]: ..nvm was looking at the wrong line 👍

**Comment by linjmeyer** [comment]
Thanks @louisjimenez, adjusted and re-requested review!

**Pull request merged**